forc-pkg/src/pkg.rs

```diff
@@ -2262,11 +2313,68 @@ pub fn build_with_options(build_options: BuildOptions) -> Result<Built> {
             let built_package = build_package_with_options(&package_manifest, build_options)?;
             Ok(Built::Package(built_package))
         }
-        ManifestFile::Workspace(_) => bail!("Workspace building is not supported"),
+        ManifestFile::Workspace(_) => {
+            bail!("Building a workspace is not supported.")
+        }
+    }
+}
+
+/// Returns a tmp PackageManifest with the members of the WorkspaceManifest inserted as
+/// dependencies
+fn tmp_manifest_with_members(
+    workspace_manifest: &WorkspaceManifestFile,
+) -> Result<PackageManifest> {
+    let mut package_manifest = PackageManifest::default();
+    let mut dependencies = BTreeMap::new();
+    let project = Project {
+        name: "worksapce".to_string(),
+        ..Default::default()
+    };
+    for (member, member_path) in workspace_manifest
+        .members()
+        .zip(workspace_manifest.member_paths()?)
+    {
+        let dependency_details = DependencyDetails {
+            path: member_path.to_str().map(|path| path.to_string()),
+            ..Default::default()
+        };
+        let dependency = Dependency::Detailed(dependency_details);
+        dependencies.insert(member.clone(), dependency);
+    }
+    package_manifest.dependencies = Some(dependencies);
+    package_manifest.project = project;
+    Ok(package_manifest)
+}
+
+/// Returns the order of workspace members for build/deploy/check.
+pub fn workspace_member_order(
+    manifest: &WorkspaceManifestFile,
+    locked: bool,
+    offline: bool,
+) -> Result<Vec<String>> {
+    let mut member_order = vec![];
+    let workspace_members: HashSet<String> = manifest.members().cloned().collect();
+    let tmp_pkg_manifest = tmp_manifest_with_members(manifest)?;
+    let tmp_pkg_manifest_file = PackageManifestFile::new(tmp_pkg_manifest, manifest.path());
+    // We are constructing the workspace wide dependency graph.
+    let workspace_build = true;
+    let build_plan = BuildPlan::from_lock_and_manifest(
+        &tmp_pkg_manifest_file,
+        locked,
+        offline,
+        workspace_build,
+    )?;
+    let graph = build_plan.graph();
+    for member in build_plan.compilation_order() {
+        let name = &graph[*member].name;
+        if workspace_members.contains(name) {
+            member_order.push(name.clone());
+        }
     }
```

Comment:
This loop could also be written as something like the following, avoiding the need for a `mut` `member_order` binding:

```rust
let member_order = build_plan
    .compilation_order()
    .map(|&n_ix| &graph[n_ix].name)
    .filter(|&name| workspace_members.contains(name))
    .cloned()
    .collect(); 
```